tests/System.Buffers.Primitives.Tests/TestHelpers.cs

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+using Xunit;
+
+namespace System.Buffers.Tests
+{
+    internal static class TestHelpers
+    {
+        public static void Validate<T>(Buffer<T> buffer, params T[] expected)
```

Comment:
You could write 2 versions of this function. One has a "where T : struct, IEquatable<T>" which handles all your value types and can be routed almost directly to Span.SequenceEqual. The other would need to do the iteration and reference comparison.